### ISSUE TITLE
[YSH] Document idiom for mutable default params (setvar param = param or default doesn't work with empty containers)

### ISSUE THREAD

**Issue by bar-g** [oilshell/oil#1842]
I see broken behavior when calling a proc/func and passing a mutable that the proc/func is supposed to modify.

In particular, the problem happens when the proc/func needs to have a mutable default. (Because mutable defaults are disallowed to avoid pointing/aliasing ambiguities, and due to the still lacking explicit `->mutable_default` syntax, this currently requires using the "out=null" workaround.)

The tests call a "modifydict" function passing the mentioned dict, but "false" means the calls fail to actually modify the specified dict.

Current results of the tests:

```
Passing global dict:

true:	 modifydict(mydict),		empty		mydict
true:	 modifydict, defaulting to	empty		mydict
false:	 modifydict(otherdict),		empty		otherdict
false:	 global mydict left untouched
true:	 modifydict(otherdict),		populated	otherdict
true:	 global mydict left untouched

Having empty globals,
passing local dict:

false:	 modifydict(mydict),		empty		mydict
false:	 modifydict, defaulting to	empty		mydict
true:	 modifydict(mydict),		populated	mydict
false:	 modifydict, defaulting to	populated	mydict
false:	 modifydict(otherdict),		empty		otherdict
true:	 modifydict(otherdict),		populated	otherdict

false:	 global mydict left untouched
true:	 global otherdict left untouched

Having unset globals,
passing local dict:

  	setvar out = out or mydict              # This (correctly) allows fall back only to global dicts.
  	                    ^~~~~~
 fatal: Undefined variable 'mydict'
```

Tests:
```

proc modifydict-default-to-mydict(; out=null) { # Unfortunately, the out=null workaround is too limited.
                                                # (See test results.)
                                                #
                                                # Todo: support explicit out->mydict (alias/pointing) defaults,
                                                # for proper call-environment based referencing,
                                                # without the "unexpected mutable default" pitfall.
                                                # While out=mutable may still produce an error until
                                                # copy-value-behavior (i.e. internal duplication of mutables)
                                                # is implemented. (Still use =null workaround to define
                                                # isolated mutables *within* the function.)

	setvar out = out or mydict              # This (correctly) allows fall back only to global dicts.

	setvar out['modifydict'] = 'done'
}


echo "
Passing global dict:
"

setvar mydict = {}

modifydict-default-to-mydict (mydict)
... echo
		"$[ mydict ===  { 'modifydict': 'done' } ]:	"
		'modifydict(mydict),		empty		mydict'
;



setvar mydict = {}

modifydict-default-to-mydict
... echo
		"$[ mydict ===  { 'modifydict': 'done' } ]:	"
		'modifydict, defaulting to	empty		mydict'
;
	
	
	
setvar mydict = {}
setvar otherdict = {}

modifydict-default-to-mydict (otherdict)
... echo
		"$[ otherdict ===  { 'modifydict': 'done' } and not mydict ]:	"
		'modifydict(otherdict),		empty		otherdict'
;
	
... echo
		"$[ mydict ===  {} ]:	"
		'global mydict left untouched'
;
	
	
	
setvar mydict = {}
setvar otherdict = { 'key': "value" }

modifydict-default-to-mydict (otherdict)
... echo
		"$[ otherdict ===  { 'key': 'value', 'modifydict': 'done' } and not mydict ]:	"
		'modifydict(otherdict),		populated	otherdict'
;
	
... echo
		"$[ mydict ===  {} ]:	"
		'global mydict left untouched'
;



proc passlocal {
	echo """
	passing local dict:
	"""

	var mydict = {}

	modifydict-default-to-mydict (mydict)
	... echo
			"$[ mydict ===  { 'modifydict': 'done' } ]:	"
			'modifydict(mydict),		empty		mydict'
	;
	
	
	setvar mydict = {}

	modifydict-default-to-mydict
	... echo
			"$[ mydict ===  { 'modifydict': 'done' } ]:	"
			'modifydict, defaulting to	empty		mydict'
	;
	
	setvar mydict = { 'key': "value" }
	
	modifydict-default-to-mydict (mydict)
	... echo
			"$[mydict ===  { 'key': 'value', 'modifydict': 'done' }]:	"
			'modifydict(mydict),		populated	mydict'
	;	
	
	setvar mydict = { 'key': "value" }
	
	modifydict-default-to-mydict
	... echo
			"$[mydict ===  { 'key': 'value', 'modifydict': 'done' }]:	"
			'modifydict, defaulting to	populated	mydict'
	;
	
	
	setvar mydict = {}
	var otherdict = {}

	modifydict-default-to-mydict (otherdict)
	... echo
			"$[ otherdict ===  { 'modifydict': 'done' } ]:	"
			'modifydict(otherdict),		empty		otherdict'
	;
	
	
	setvar mydict = {}
	setvar otherdict = { 'key': "value" }

	modifydict-default-to-mydict (otherdict)
	... echo
			"$[ otherdict ===  { 'key': 'value', 'modifydict': 'done' } ]:	"
			'modifydict(otherdict),		populated	otherdict'
	;
}


echo "
Having empty globals,"

setvar mydict = {}
setvar otherdict = {}

passlocal

echo
... echo
		"$[ mydict ===  {} ]:	"
		'global mydict left untouched'
;

... echo
		"$[ otherdict ===  {} ]:	"
		'global otherdict left untouched'
;


echo "
Having unset globals,"

unset mydict
unset otherdict

passlocal

if test -z "${mydict+isset}"
then
   echo 'true:	global mydict left untouched'
else
   echo 'false:	global mydict left untouched'
fi

if test -z "${otherdict+isset}"
then
   echo 'true:	global otherdict left untouched'
else
   echo 'false:	global otherdict left untouched'
fi

```

**Comment by bar-g** [comment]
Result of the tests:  
```
Passing global dict:

true:	 modifydict(mydict),		empty		mydict
true:	 modifydict, defaulting to	empty		mydict
false:	 modifydict(otherdict),		empty		otherdict
false:	 global mydict not falsely modified
true:	 modifydict(otherdict),		populated	otherdict
true:	 global mydict not falsely modified

Having empty globals,
passing local dict:

false:	 modifydict(mydict),		empty		mydict
false:	 modifydict, defaulting to	empty		mydict
true:	 modifydict(mydict),		populated	mydict
false:	 modifydict, defaulting to	populated	mydict
false:	 modifydict(otherdict),		empty		otherdict
true:	 modifydict(otherdict),		populated	otherdict

false:	 global mydict not falsely modified
true:	 global otherdict not falsely modified

Having unset globals,
passing local dict:

  	setvar out = out or mydict              # This (correctly) allows fall back only to global dicts.
  	                    ^~~~~~
 fatal: Undefined variable 'mydict'
```

**Comment by andychu** [comment]
I don't understand what this is meant to show -- can you summarize the problem you're seeing?

**Comment by bar-g** [comment]
> I don't understand what this is meant to show -- can you summarize the problem you're seeing?

I added an intro to the issue summary https://github.com/oilshell/oil/issues/1842 , is it understandable now?

**Comment by andychu** [comment]
Ohh yess, Hmmmm  

Sorry `setvar out = out or mydict` has a bit of a pitfall -- when `out` is the **empty** dict.

The `or` operator tests if a value is "falsey", and that includes empty  containers -- `{} [] false null`

This is a bit of an unfortunate interaction -- it's something Python also has.  I actually tried to change what the `or` operator meant, but it causes other problems elsewhere in the language.

----

So really the right  thing for dicts is

```
if (param === null) {
   setvar param = defaults
}
```

rather than

```
setvar param = param or defaults
```

**Comment by andychu** [comment]
Shell also has a related issue, it has both `${a:-default}` and it has `${a-default}`, for the difference between empty and unset

---

It may be possible to introduce some kind of other operator, but I think it shouldn't be `||` as it is in JavaScript

```
setvar param = param || defaults  # JS style, but this could be confused with shell ||
```

Or perhaps

```
setvar param = param ? defaults   # ? could be the "test for null operator"
```

----

There is a related thing about perhaps about `null` attribute chaining, some languages have this

```
if (mydict.?myattr.?otherattr) {
}
```

That's sort of like `mydict => get('myattr')`, but that only works for one level

Hm

---

I think this is the reason we would need to document the idiom, but yeah it's not quite trivial

**Comment by andychu** [comment]
Probably the simplest solution is just to implement `deepcopy()` on objects. 

It won't require any new syntax, and then you can just them as a default arg

```
typed proc p (; ; mydict={}) {
}
```

Although there is also a rare possibility that you **don't** want deepcopy() -- you want to mutate the same thing!

So it's kinda tough

**Comment by bar-g** [comment]
Yeah, actually for my "swimming lane" example (if you've seen it) I think I'd not want deepcopy() but pointer/alias behavior, because it's supposed to modify the default (to propagate changes).

---
> typed proc p (; ; mydict={})

  ^ typed? that's new to me

---

The improved workaround "idom" surprised me.

I re-ran the tests with this change:

	#setvar out = out or mydict 
	if (out === null) {
   	setvar out = mydict
   	
	}

and it actually allowed passing an empty (non-default) "otherglobal". So, this is already better.

However, passing local dicts still fails. I think this is because they are are not accessible from within the called proc.

I guess the local dict would only be accessible on the proc call-site, and thus also weigh-in for the need of the `out->localdict` syntax to allow referencing (external) mutable defaults.

As the tests show, unfortunately the problem seems not quite solved yet.

**Comment by bar-g** [comment]
Hm, in terms of a correct syntax, these tests would probably actually be one of the rare cases of using `out->:somedict` here (`->:`to explicitly refer to a mutable and not to a Place).

**Comment by bar-g** [comment]
(from [1863](https://github.com/oilshell/oil/issues/1863#issuecomment-2000081770))
> OK I still don't understand what the problem is there... there's too much code, and it's not in context.
>
> What did you expect to happen, and what happened? For a specific case

I expect the test to return true, and already tried to explain what I did, what happend, and what I suspect:

(It's no problem if you missed it. And no wonder if being overworked, but if that's the case I'd rather say to get some healthy recreational time.)

--- 

I re-ran the tests with this change:

    #setvar out = out or mydict 
    if (out === null) {
       setvar out = mydict
    }

and it actually allowed passing an empty (non-default) "otherglobal". So, this is already better.

[ The first two tests now showed:
 true:	 modifydict(otherdict),		empty		otherdict
 true:	 global mydict left untouched
]

However, passing local dicts still fails. I think this is because they are are not accessible from within the called proc.

I guess the local dict would only be accessible on the proc call-site, and thus also weigh-in for the 
need of the `out->localdict` syntax to allow referencing (external) mutable defaults.

---
 
I wouldn't know a better solution than:

```
proc modifydict-default-to-mydict(; out->mydict) { 
                                                # That is,
                                                # support explicit out->mydict (alias/pointing) defaults,
                                                # for proper call-environment based referencing,
                                                # without the "unexpected mutable default" pitfall.
                                                # While out=mutable may still produce an error until
                                                # copy-value-behavior (i.e. internal duplication of mutables)
                                                # is implemented. (Still use =null workaround to define
                                                # isolated mutables *within* the function.)

	set-> out['modifydict'] = 'done'
}

```

**Comment by andychu** [comment]
> However, passing local dicts still fails. I think this is because they are are not accessible from within the called proc.

OK yeah I don't know what you mean by this

I would need an isolated example

**Comment by andychu** [comment]
Also in general I would say it's mostly a bad idea to go back and edit the first comment.  I might not have read the comment to begin with, so if you go edit it, it's doubly-confusing.

It's better just to copy and paste the whole thing, and then **omit** the parts that aren't relevant, and produce a **reduced** test case

https://gcc.gnu.org/bugs/minimize.html

https://webkit.org/test-case-reduction/

**Comment by bar-g** [comment]
Hm, but it is only a single trivial proc that gets repeatedly called in different ways (and the proper result is tested), first in global scope, then from within another proc.

If I convert this to multiple "reduced" cases the code "explodes" and we would get a multitude of separate issues filed here. I believe it's much easier to take a minute to understand the testing sequence and have all the cases in one file. The output is also providing some contextual overview.

**Comment by bar-g** [comment]
But to summarize, my conclusion from these tests is that the workaround (old and new) does not work and the only proper solution I can imagine is the `out->mydict` syntax for defaults.

In the meantime, it could also work to allow `out=mydict`, i.e. mutable defaults. They are not always pitfalls, but also a legit use case. (Although with ugly (i.e. misleading) `=` syntax.)

**Comment by andychu** [comment]
I think this is the same bug as

-  #1841 

When we fix that, it should all go away.  If not, we can revisit then

**Comment by bar-g** [comment]
Where do you see a call of `setglobal`?

**Comment by bar-g** [comment]
> passing local dicts still fails. I think this is because they are are not accessible from within the called proc

To be more precise, this is what I based my conclusion on: I think they would only be accessible at the proc call-site, thus out->mydict, but I don't know when exactly the proc definition-site is evaluated.

**Comment by andychu** [comment]
The proc definition is currently always evaluated globally.  We disallow nested procs for this reason (this may change)

I think maybe the confusion is that you can't define a default var inside the body of a proc.

The defaults are always constant expressions like `42` or global like `mydict`.  That's it

**Comment by bar-g** [comment]
> maybe the confusion is that there's no such thing as a "local default"

Hm, from within a proc it's not possible to access (or default to) a local variable from a calling proc, understood.

However, I think it's important for a proc to be able to default to any of the variables available in the calling context. It's pretty incomplete, otherwise.



Or/And, is what you were NOT telling me, that the calling proc must instead deal with assigning, defaulting, and pass a separate variable?

    var use_this = myvar or defaultvar; p (use_this)

That would also look to me like a pretty hard dent for procs that are supposed to be contained and universally usable.



> The proc definition is currently always evaluated globally. We disallow nested procs for this reason (this may change)

Ok, sorry I think the defaults are a point towards _on-call_ in-context _initialisation_. (Maybe even independent of allowing nested procs or not.)

**Comment by andychu** [comment]
Yeah no, that doesn't exist.  There is no dynamic scope in YSH

It's not a problem with the language, because people have written millions and millions of lines of Python and JavaScript without this behavior.

This is why it's important for you to explain in the bug reports WHAT YOU EXPECT to happen.  

Do NOT merely assert that whatever you decided to mark as "false" is a bug.  

It MIGHT be a bug, or it might be that you misunderstood the language.

Both have happened, so you have to explain what your expectation is.

**Comment by andychu** [comment]
Closing in favor of #1866 

Again, please file bugs with minimized test cases

**Comment by bar-g** [comment]
> It MIGHT be a bug, or it might be that you misunderstood the language.
> Both have happened,

Yes, definitely.

Here, I didn't subsume it under dynamic scope (which I was placing within procs and their nesting), but as the interface between the calling and the called scope, and expected the capability of the interface (proc definition) to default to to any outer state.

With the other issues that the tests here bought up now triaged or solved (the working workaround), I now also think the larger obstacles have been moved out of the way.

So thank you also for answering my proc evaluation question.

---

> you have to explain what your expectation is.

What I expect (after ironing things out) is "a smooth upgrade experience from shell, and pitfalls removed".

A couple of my rather vague "not working" issues filed came from that, me not being able to exactly point to the reason of the problem nor a solution. I will try to better explain goals and expectations next time.

Partly, I noticed that some answers I got were a bit unfortunate, not helping me much, but already in the direction of the expectations. For example at the issue of detecting undefined variables in ysh without geting fatal errors, I unfortunately just didn't realize at first that `test ${var+isset}` would continue to do it also for typed vars.

I agree that on many occasions, things have shown not to be a problem with the language, but yes with error messages, or the docs.

That's why I had tried to merge three or five words into the "troubleshooting" faq, to tie in missing info sure to cause problems at some time if not known. (`write` is not generally equivalent to `echo` in it's default operation, and with the examples glossing over it, there will be users getting errors that look like output behaving inconsistently once lists contain multiple entries or similar. I think that's a quite an un-smooth pitfall lurking in the middle of ysh.)

Inconsistent variable interaction I think is another, and completely unnecessary (python inherited) ysh pitfall. It's a drag for everybody not accustomed and trained daily to continuously having to think about a mutable vs atoms dichotomy without there being any syntax difference, and thus also no specific error messages possible, and no syntax reminders being apparent anywhere. Especially relevant of course for an only occasional, and mere shell scripting user.

The absence of the dynamic scope, I think is rather something that could be dealt with nicely by some strict_ error messages, and good hints in the shell-versus-ysh doc, i.e. comparing general programing patterns.

**Comment by bar-g** [comment]
Oh, some topics like places and rebind, these are things I also subsume as solvable with a consistent `->` syntax, i.e. a `strict_type_interaction` design touching maybe 7-10 places, instead of only 4-5.

**Comment by bar-g** [comment]
> It MIGHT be a bug, or it might be that you misunderstood the language.

It seems important, though, to note that it can be both. The misunderstanding by shell users resulting from problems of the YSH language.

Like when proc calling semantics change depend on passed type, and what shell user would expect proc defaults to only be initalized once at parse time?

I'll try to make the expectations more clear in the corresponding issues.
